internal/datacoord/stats_task_meta.go

```diff
@@ -295,10 +269,72 @@ func (stm *statsTaskMeta) CanCleanedTasks() []int64 {
 
 	needCleanedTaskIDs := make([]int64, 0)
 	for taskID, t := range stm.tasks {
-		if t.GetState() == indexpb.JobState_JobStateFinished ||
-			t.GetState() == indexpb.JobState_JobStateFailed {
+		if t.GetCanRecycle() && (t.GetState() == indexpb.JobState_JobStateFinished ||
+			t.GetState() == indexpb.JobState_JobStateFailed) {
 			needCleanedTaskIDs = append(needCleanedTaskIDs, taskID)
 		}
 	}
 	return needCleanedTaskIDs
 }
+
+func (stm *statsTaskMeta) GetAllTasks() map[int64]*indexpb.StatsTask {
+	stm.RLock()
+	defer stm.RUnlock()
+
+	return stm.tasks
```

Comment:
Return this is not safe, there still may occur data race since another gorountine may direct read stm.task which the RLock is unhold.